### ISSUE TITLE
Properly collecting items

### ISSUE THREAD

**Issue by iProgramMC** [GrowtopiaNoobs/GrowtopiaServer#976]
I have playingo's sendTake, however it feels kind of weird because the dropped item's ID is different for each client. For example, client 1's item UID could be 0xfc24a9b0, whilst client 2's item UID (which is referring to the same item) could be 0x809a2bd7.

Could anyone help with this?

**Comment by NotDark** [comment]
i guess u need set default id in sendworld

**Comment by iProgramMC** [comment]
> i guess u need set default id in sendworld

wdym

**Comment by MarmoriPelaa** [comment]
I've been trying to implement this too, from scratch. And I'm facing the exact same issue. I wonder if playingo could help?

**Comment by iProgramMC** [comment]
I know Playingo on Discord. He hasn't come online in a while.

I joined his GTV3 development server and have helped him.

**Comment by kingawesomed2** [comment]
@iProgramMC how to change item color

**Comment by iProgramMC** [comment]
> @iProgramMC how to change item color

first of all, wdym
second, create another issue

**Comment by Lapada** [comment]
i have no idea why it doesn't work, playingo is the best coder...

**Comment by iProgramMC** [comment]
> i have no idea why it doesn't work, playingo is the best coder...
@Lapada This is a serious thread asking for help. It works on client side but for other clients it DOESN'T.

**Comment by MarmoriPelaa** [comment]
I think NotDark is right..
As you might know, if you allocate some memory, it just takes a pointer to a location that is currently marked as free space. This means that when you allocate space for the "default id", there might be some trash data which you have to nullify. In order to do this, you would have to debug what you send:
1. make a for loop that prints out every byte in hex (e.g. with `printf("%02hhx ", raw[i]);` where raw would be the full packet that includes the world data)
2. drop an item on the client
3. collect the item on the client
4. look for the `plantingTree` value that came from collecting item packet (which is the netID for the item) in the world sending packet
5. if you found the `plantingTree` in the packet, then check the offset of `plantingTree` in the packet
6. change the value at the offset of the world packet to 0 in code.

Example:
```Cpp
for(int i = 0; i < allocMem; i++) {
    printf("%02hhx ", data[i]);
}
```
^-- put that at the end of `sendWorld(...)` before the `delete data;`. This is for printing the world data packet.
In sendTake, print the plantingTree value. I don't know what it looks like for you but this is what I would use:
```Cpp
void sendCollect(ENetPeer* peer, int netID, int itemNetID) {
    //...
    cout << std::to_string(itemNetID); //plantingTree of the received collect packet
    //...
}
```

I'm going to try this for myself now. I am really tired of floating items.

**Comment by iProgramMC** [comment]
I will also try this for myself, thank you for the help @MarmoriPelaa

**Comment by iProgramMC** [comment]
@MarmoriPelaa I've done it **almost** the same way as you did, but I wrote the data to a file. I used FC to compare the files, and there was no difference.

**Comment by iProgramMC** [comment]
@MarmoriPelaa do you have discord?

**Comment by iProgramMC** [comment]
> the sendtake ive sent was an old one and in order to fix the collecting u will sadly have to "fix" the entering world packet completely, as it isnt properly done as how anybody made it in the source. the packet isnt "fully reversed" but lapada also told me they fully reversed the packet in order to use it properly which should fix the collecting issues at the same time.
> plantingTree is basically the "netID" of the collected item and can even be modified depends on how you use the sendworld packet, and you got to use it differently than anybody's default packet according to lapada

If you don't mind, could you please give me the fixed version? If not, can you at least explain how to change it? 

Yeah, the sendWorld is kinda bad, not only that, but the weather also changes from time to time (though mostly sunny)

**Comment by playingo** [comment]
exactly, i know this already. tbh i dont have the fixed version of the sendWorld im just using some kind of "trick" so the other clients have the "same" id but its not proper, as worlds that have less than 5 characters forexample a world called ("1234") will nearly always not work with collecting as the ids are just completely random for every client.... Also yes sometimes its really changing the weather, ive noticed this too a long time ago, and its because of the broken sendWorld packet as i said.

**Comment by GrowtopiaNoobs** [comment]
@playingo Yes, you are correct, but it is about missing data. I will release new source sometime, but don't except it anytime soon.

**Comment by playingo** [comment]
thats good, also i think some crash fixes should be added as default in the source maybe, if you want i could give you some but not too many as i dont have too many aswell.

**Comment by iProgramMC** [comment]
> thats good, also i think some crash fixes should be added as default in the source maybe, if you want i could give you some but not too many as i dont have too many aswell. also i would add the TileExtra (or "BlockVisuals") how some people call it as standard in the source so anyone has the chance to find out new stuff with it.

i dm'd you on discord

**Comment by playingo** [comment]
yeah u know why  i cant be online 24/7 ive told u alr

**Comment by playingo** [comment]
UPDATE @MarmoriPelaa @GrowtopiaNoobs we've just achieved to repair the whole sendWorld function, and it now does NO LONGER have any missing data (except dropped items *YET*) but data like weather when entering (without the use of OnSetCurrentWeather) and the fix of the dropped item uid. Thanks for all who have been involved in this issue thread (except lapada)!

**Comment by iProgramMC** [comment]
oh nice :D

**Comment by MarmoriPelaa** [comment]
Sounds good. Did you make a PR or where can I find the fixed packet?

**Comment by iProgramMC** [comment]
I didn't make a PR because I promised to Playingo I wouldn't reveal it, but what I can do is give a hint.

Basically what the game does is it reads over the sent packet's length (so reads garbage data). That's the reason. You will need to provide a modified send world packet.

**Comment by playingo** [comment]
@MarmoriPelaa its going to be available soon, but its currently used only for gtv3 to avoid competition. it's not a very difficult fix though.

**Comment by Lapada** [comment]
translated: we want help from others but hehe we won't help you if we find out !! we just like to flex with our 30 second fixes

**Comment by playingo** [comment]
@Lapada i mean i am willing to tell marmoripelaa but not in public here because i dont want everyone who doesnt work hard to achieve something to know how to fix it. @MarmoriPelaa incase you want the "fix" (as i believe you got the rights to, aswell as @GrowtopiaNoobs), i would like to tell you on my discord DiruX#4989.
we want that those who dont work hard shouldnt get the fix, those who do / were involved in this thread before this comment may have the fix. Thanks

**Comment by iProgramMC** [comment]
> @Lapada i mean i am willing to tell marmoripelaa but not in public here because i dont want everyone who doesnt work hard to achieve something to know how to fix it. @MarmoriPelaa incase you want the "fix" (as i believe you got the rights to, aswell as @GrowtopiaNoobs), i would like to tell you on my discord DiruX#4989.
> we want that those who dont work hard shouldnt get the fix, those who do / were involved in this thread before this comment may have the fix. Thanks

you dont even remember your discord hehe
that name doesn't work so should I reveal your actual discord @playingo

**Comment by Lapada** [comment]
@playingo took you a few days to think of an actual response

**Comment by coderfixerios** [comment]
**i dont know how to do this bruh
**
